forecasting/src/autogluon/forecasting/utils/dataset_utils.py

```diff
@@ -0,0 +1,297 @@
+    1            C           1           2           3
+    2            B           1           2           3
+
+    """
+    if index_column is None:
+        X = X[[time_column, target_column]]
+        X["index_column"] = ["time_series" for i in range(X.shape[0])]
+        index_column = "index_column"
+    time_list = sorted(list(set(X[time_column])))
+    freq = pd.infer_freq(time_list)
+    # check for uniform date
+    if freq is None:
+        raise ValueError("Freq cannot be inferred. Check your dataset.")
+
+    def reshape_dataframe(df):
+        """
+        for each time occurs in the dataset, we select the target value corresponding to
+        each time series, and use dataframe.pivot() to convert it to one column, where the column name is the
+        time, each row is the corresponding target value for each time series.
+        """
+        df = df.sort_values(by=index_column)
+        data_dic = {index_column: sorted(list(set(df[index_column])))}
+        # check for dataset with multiple targets for a certain time series/time
+        if any(df[[index_column, time_column]].value_counts() > 1):
+            for combination in df[[index_column, time_column]].value_counts().index:
+                raise ValueError(f"Containing multiple targets for time series {combination[0]} and time {combination[1]}. "
+                                 "Please check your dataset.")
+        # check whether need auto-padding
+        need_padding = False
+        if any(df[time_column].value_counts() != len(df[index_column].unique())):
+            logger.log(30, "Warning: autogluon requires each index to be observed over the same set of time values. \n"
+                           "As this is not the case in your data, we are automatically padding the dataset with all missing (index, time) combinations which may take some time. \n"
+                           "To do this yourself, simply add rows to the Dataframe with target value = NA for each missing (index,time) combination.")
+            need_padding = True
+
+        for time in time_list:
+            tmp = df[df[time_column] == time][[index_column, time_column, target_column]]
+            tmp = tmp.pivot(index=index_column, columns=time_column, values=target_column)
+            # automatically padding with NAN if for some time series missing a date target if needed
+            if need_padding:
+                for index in data_dic[index_column]:
+                    if index not in tmp.index:
+                        tmp.loc[index, time] = None
+            tmp_values = tmp[time].values
+            data_dic[time] = tmp_values
+        return pd.DataFrame(data_dic)
+
+    X = reshape_dataframe(X)
+    return X
+
+# TODO: Improve the way to do the split
+def train_test_split_dataframe(data, prediction_length):
+    test_ds = data.copy()
+    train_ds = data.iloc[:, :-prediction_length]
+    if not any(data.iloc[:, -prediction_length:]):
```

Comment:
this is checking the test set right?

Reply:
> this is checking the test set right?

The actual data in the test dataset is the last `prediction_length` data points of each time series. So this is checking whether the last `prediction_length` data points of each time series contain only NAN.

Reply:
Yes we may want to check this logic a bit. We should ignore any time series with any `NaN` in the backtest window or at least a large portion since we can't compute metrics on time series with NaNs.
I think it would be helpful to set this to `eval_window_test_ts = data.iloc[:, -prediction_length:]` or we can call it `backtest_ts` but let's think about how to generally extend to multi backtest.

Reply:
I'll add a todo for this, but thanks for raising this, I believe it is a bug. 
```
df = pd.DataFrame({
    "a": np.array([np.nan, np.nan, np.nan]),
    "b": np.array([np.nan, np.nan, np.nan]),
})

assert any(df)
```